lib/rucio/daemons/rsedecommissioner/decommission_rse.py

```diff
@@ -0,0 +1,107 @@
+) -> None:
+    """RSE decommissioning template function.
+
+    RSE decommissioning proceeds in common steps of
+    - Profile initialization
+    - Discovery of rules to act (move / delete / etc.) on
+    - Actual handling of the rules
+    - Finalization
+    This function takes the functions corresponding to the four steps as its arguments and
+    executes them in order.
+
+    The handlers are given as a list of pairs of functions. The two functions both have a
+    signature
+    (rule, rse, *, logger=logging.log) -> bool
+    with the following actions:
+    - The first function is the "condition" against which each rule is tested. The returned
+      boolean indicates whether the rule satisfies it.
+    - The second function is the "(in)action" taken against the rule if the first function
+      returns True. The returned boolean indicates whether the limit checker should be
+      incremented after executing the action.
+
+    For each rule listed by the discoverer function, the conditions are applied in the order
+    given in the handlers list until one evaluates to True, at which point the corresponding
+    action function is executed. All remaining condition-action pairs are skipped for the rule.
+
+    :param rse: RSE table entry as a dictionary.
+    :param profile: Decommissioning profile.
+    :param logger: Logging function.
+    """
+    max_rules = config_get_int('rse-decommissioner', 'max_rules', default=-1)
+    max_locks = config_get_int('rse-decommissioner', 'max_locks', default=-1)
+    if max_locks < 0 and max_rules < 0:
+        raise RucioException('This daemon requires configuration parameters'
+                             ' rse-decommissioner/max_rules and rse-decommissioner/max_locks'
+                             ' set to non-negative integers.')
+
+    # Initialize the RSE for decommissioning.
+    profile.initialize(logger=logger)
+    # Set up throttling.
+    limit_check = LimitChecker(max_rules, max_locks)
+
+    # Run over rules locking datasets / replicas at the RSE.
+    try:
+        rules = profile.discover(logger=logger)
+        for rule in rules:
+            result = profile.process(rule, logger=logger)
+            if result is not None:
+                limit_check.processed_rules += 1
+                if result:
+                    # Rule removed; increment the counters and check the throttling conditions.
+                    # increment() raises DeletionLimitExceeded when a counter exceeds set maximum.
+                    limit_check.increment(rule, logger=logger)
+    except DeletionLimitExceeded as exc:
+        logger(logging.INFO, '(%s) Stopping decommissioning cycle because %s', rse['rse'], exc)
+
+    if limit_check.deleted_rules > 0:
+        logger(logging.INFO,
+               '(%s) %s rules are being processed. Decommissioning is not complete.',
+               rse['rse'], limit_check.deleted_rules)
+    elif limit_check.processed_rules != 0:
```

Comment:
You’ll have to forgive me, I’m not confident I can follow the logic in this particular sequence.

If `result is None`, then none of the handlers matched the rule. If this happens for all rules on an RSE, then at the end `deleted_rules == 0` and `processed_rules == 0`. But, of course, the RSE isn’t actually empty. Is this accurate or am I missing something?

PS: Maybe the mistake is that the generic `finalize()` function will claim that there are no rules. But, after reviewing the entire PR, this is the only portion that makes me somewhat uneasy. I can’t simulate it in my head with confidence.

Reply:
Indeed, things have gotten mixed up around here. I straightened up the logic. Now we have four counters:

- `processed_rules`: Anything that `profile.process()` was called up on.
- `handled_rules`: Rules for which `result` was not None (matched some condition).
- `deleted['rules']`: Rules for which `result` was true (matched some condition and the action was to delete).
- `deleted['locks']`: Number of deleted locks.

Then the logic and messaging are
- If `deleted['rules'] > 0`, say that there are rules being deleted. Finish the decommissioning cycle.
- If `handled_rules > 0`, say that nothing was deleted but there are rules expected to disappear soon. Finish the cycle.
- If `processed_rules > 0`, say that there were rules that were not handled by the current process. Set the decommissioning status to SUSPENDED.
- Otherwise: try finalizing the RSE.

Reply:
The logic around this code has evolved quite a lot over the time, but I think it's in a good shape now. In the current form, there is actually no need for either a separate throttling counter class or an exception to be thrown just to get out of an un-nested loop, so I got rid of throttling.py and just started using plain counter variables within `decommission_rse()`.